### DISCUSSION TITLE
Adding support for tag_label in IPC module and deprecating tag_output

### DISCUSSION THREAD

**Discussion by madhavpcm** [polybar/polybar#2840] [Ideas]
TAG_OUTPUT is deprecated in other modules and uses TAG_LABEL instead. IPC module doesnt support TAG_LABEL and I would like to extend its use to the same.

**Comment by patrick96** [comment]
Go for it!

**Comment by madhavpcm** [comment]
When Im trying to call load_label() in the constructor of ipc module, im getting a null value. Is there any reason why? Im following a similar procedure as in text module.

**Reply by patrick96** [reply]
It's impossible for `load_label` to return a nullptr. It either throws an exception or creates some label. Most likely you are not actually calling load_label because it is behind some if check

**Reply by madhavpcm** [reply]
> It's impossible for `load_label` to return a nullptr. It either throws an exception or creates some label. Most likely you are not actually calling load_label because it is behind some if check

my bad, the return value of load_label seems to be 0. And then the next exception I get is Mod
```polybar|error: Disabling module "demo" (reason: Missing parameter "module/demo.label")```

**Reply by patrick96** [reply]
That's also not possible, `load_label` returns a `shared_ptr` and not an integer. The exception looks like one that would be thrown inside `load_label` when it tries to load the label contents when the label is not actually defined in the config.

**Reply by madhavpcm** [reply]
> That's also not possible, `load_label` returns a `shared_ptr` and not an integer. The exception looks like one that would be thrown inside `load_label` when it tries to load the label contents when the label is not actually defined in the config.

Ok, But in the case the text modules defined in my config do not have label defined.

 Also how should the values inside m_label updated when we  change the output as in this context
```cpp

  void ipc_module::exec_hook() {
    // Clear the output in case the command produces no output

    m_label->clear();
    try {
      command<output_policy::REDIRECTED> cmd(m_log, m_hooks[m_current_hook]->command);
      cmd.exec(false);
      cmd.tail([this](string line) { m_label.set(line) ; }); // ?? 
    } catch (const exception& err) {
      m_log.err("%s: Failed to execute hook command (err: %s)", name(), err.what());
      m_label->clear();
    }

    broadcast();
  }

```
should I create setter for m_label or is there some other way possible ?

Is there any documentation for m_tokenizer I can read up on ? Im not sure I grasp its control flow correctly.

**Reply by patrick96** [reply]
> Ok, But in the case the text modules defined in my config do not have label defined.

Yes. The text module only loads a label if you actually use the `<label>` tag in `format`.

Labels are updated by giving their tokens a value, this happens in the `update` method.
First you reset all tokens from the last time using `reset_tokens` and then for each token call `replace_token` on the label. Here is how this looks like in the date module:

```cpp
    if (m_label) {
      m_label->reset_tokens();
      m_label->replace_token("%date%", m_date);
      m_label->replace_token("%time%", m_time);
    }
```

In the ipc module, the label should probably have an `%output%` token, just like in the script module.

After replacing the tokens, you have to call `broadcast` to inform the controller that the module has changed (you can also remove the `broadcast` call from the `exec_hook` function).

**Comment by madhavpcm** [comment]
Hey,
I have got something working, 
```
[module/demo]
type = custom/ipc
label= %output%
hook-0 = echo foobar
hook-1 = date +%s
hook-2 = whoami
format= <label>
format-background = #ff2
format-0 = <label>
format-1 = 
format-2 = <label>

; Based on colors defined by user
format-0-background= ${color.shade1}
format-2-background= ${color.shade3}
initial = 1
click-left = "#demo.hook.0"
click-right = "#demo.hook.1"
double-click-left = "#demo.hook.2"
```
for this particular config, since format-1 is not defined, there is no output from the module and it vanishes. Should I fallback with label for the empty format with label_failed or some other label ?
Im not sure whether my overall implementation is using label or m_output, so could you check if my approach is right ?

**Reply by patrick96** [reply]
If the user explicitly set a format to the empty string, they probably want the module to vanish in this case. I don't think having a fallback for that makes sense, it's also not something any other module does.
